analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,241 @@
+
+            foreach (var routeParam in routeParams)
+            {
+                routeParameters
+                    .GetOrAdd(routeParam.Parts[0], _ => new List<RouteParameter>(1))
+                    .Add(new(routeParam.Parts.Length == 2 ? routeParam.Parts[1] : ImplicitStringConstraint,
+                        route,
+                        CalculateRouteParamLocation(routeNode.GetLocation(), routeNode, routeParam.Param)));
+            }
+        }
+
+        return routeParameters;
+    }
+
+    private static Location CalculateRouteParamLocation(Location attributeLocation, AttributeSyntax routeNode, string routeParam)
+    {
+        if (GeneratedCodeRecognizer.IsRazorGeneratedFile(attributeLocation.SourceTree))
+        {
+            return null;
+        }
+
+        return routeNode.ArgumentList.Arguments[0].Expression.RemoveParentheses() switch
+        {
+            var expression when expression.ToString().IndexOf(routeParam, StringComparison.InvariantCulture) >= 0 => CreateLocationFromRoute(expression.GetLocation(), expression.ToString()),
+            var expression => expression.GetLocation()
+        };
+
+        Location CreateLocationFromRoute(Location routeLocation, string route)
+        {
+            return Location.Create(routeLocation.SourceTree, new TextSpan(routeLocation.SourceSpan.Start + route.IndexOf(routeParam, StringComparison.InvariantCulture), routeParam.Length));
+        }
+    }
+
+    private sealed class PropertyTypeMismatch
+    {
+        public TypeSyntax Type { get; }
+        public string ConstraintType { get; }
+        public string Route { get; }
+        public Location RouteParamLocation { get; }
+
+        private bool IsImplicitStringConstraint => ConstraintType.Equals("string", StringComparison.OrdinalIgnoreCase);
+        private bool CanReportSecondaryLocation => RouteParamLocation is not null;
+
+        public PropertyTypeMismatch(TypeSyntax type, string constraintType, string route, Location routeParamLocation)
+        {
+            Type = type;
+            ConstraintType = constraintType;
+            Route = route;
+            RouteParamLocation = routeParamLocation;
+        }
+
+        public string ToPrimaryMessage()
+        {
+            if (CanReportSecondaryLocation)
+            {
+                return string.Format(MessageWithSecondaryLocationFormat, GetTypeName(Type));
+            }
+
+            if (IsImplicitStringConstraint)
+            {
+                return string.Format(ImplicitStringPrimaryMessageFormat, GetTypeName(Type), Route);
+            }
+
+            return string.Format(PrimaryMessageFormat, GetTypeName(Type), ConstraintType.ToLower(), Route);
+        }
+
+        private static string GetTypeName(TypeSyntax type) =>
+            type switch
+            {
+                ArrayTypeSyntax _ => type.ToString(),
+                PointerTypeSyntax _ => type.ToString(),
+                _ => type.GetName()
```

Comment:
This is the only way I found to keep the `[]` and `*` in the type name without doing it manually.
Otherwise, with `GetName`, I would always end with the underlying type.
For instance:
- `int[]` becomes `int`
- `int*` becomes `int`

Using the `ToString` was the only way I found it. Based on my test, the results were expected, but I might have missed a test case.

If you know a better approach, I am all ears.